src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/StorePal.LoaderPal.cs

```diff
@@ -56,5 +56,74 @@ public void MoveTo(X509Certificate2Collection collection)
                 }
             }
         }
+
+        private sealed class ApplePkcs12CertLoader : ILoaderPal
+        {
+            private readonly ApplePkcs12Reader _pkcs12;
+            private readonly SafeKeychainHandle _keychain;
+            private readonly SafePasswordHandle _password;
+            private readonly bool _exportable;
+
+            public ApplePkcs12CertLoader(
+                ApplePkcs12Reader pkcs12,
+                SafeKeychainHandle keychain,
+                SafePasswordHandle password,
+                bool exportable)
+            {
+                _pkcs12 = pkcs12;
+                _keychain = keychain;
+                _exportable = exportable;
+
+                bool addedRef = false;
+                password.DangerousAddRef(ref addedRef);
+                _password = password;
+            }
+
+            public void Dispose()
+            {
+                _pkcs12.Dispose();
+
+                // Only dispose the keychain if it's a temporary handle.
+                (_keychain as SafeTemporaryKeychainHandle)?.Dispose();
```

Comment:
should you also add a check (i.e. Debug.Assert) checking if `_keychain` is disposable but is not SafeTemporaryKeychainHandle?

Reply:
I don't know what you mean.  Keychain is always disposable.  If it's a PersistKeySet import we'll import straight into the default keychain (login.keychain) which is a SafeKeychainHandle, but not a SafeTemporaryKeychainHandle

Reply:
I'm just confused that if `_keychain` is somehow wrong type (now or in the future) this will be no op if type doesn't match

Reply:
I'm just confused that if `_keychain` is somehow wrong type (now or in the future) this will be no op if type doesn't match

Reply:
There are currently only two keychain types: `SafeKeychainHandle` and `SafeTemporaryKeychainHandle : SafeKeychainHandle`.  The comment explains that we only want to dispose temporary handles, and the code does that.

If there's some sort of third kind of keychain handle added in the future, we'll have to deal with that as appropriate, but I can't really imagine anything other than "persisted" and "temporary"... so I think we're pretty well future-proofed. (Famous last words?)

FWIW, this is just moved code.

Reply:
> FWIW, this is just moved code.

fair enough